src/analysis/normed_space/lp_equiv.lean

```diff
@@ -0,0 +1,172 @@
+space, the bounded continuous functions `α →ᵇ β` correspond exactly to `lp (λ _, β) ∞`. Here there
+can be more structure, including ring and algebra structures, and we implement these equivalences
+accordingly as well.
+
+We keep this as a separate file so that the various $$L^p$$ space files don't import the others.
+
+Recall that `pi_Lp` is just a type synonym for `Π i, E i` but given a different metric and norm
+structure, although the topological, uniform and bornological structures coincide definitionally.
+These structures are only defined on `pi_Lp` for `fintype α`, so there are no issues of convergence
+to consider.
+
+While `pre_lp` is also a type synonym for `Π i, E i`, it allows for infinite index types. On this
+type there is a predicate `mem_ℓp` which says that the relevant `p`-norm is finite and `lp E p` is
+the subtype of `pre_lp` satisfying `mem_ℓp`.
+
+## TODO
+
+* Equivalence between `lp` and `measure_theory.Lp`, for `f : α → E` (i.e., functions rather than
+  pi-types) and the counting measure on `α`
+
+-/
+
+open_locale ennreal
+
+section lp_pi_Lp
+
+variables {α : Type*} {E : α → Type*} [Π i, normed_add_comm_group (E i)] {p : ℝ≥0∞}
+
+/-- When `α` is `finite`, every `f : pre_lp E p` satisfies `mem_ℓp f p`. -/
+lemma mem_ℓp.all [finite α] (f : Π i, E i) : mem_ℓp f p :=
+begin
+  rcases p.trichotomy with (rfl | rfl | h),
+  { exact mem_ℓp_zero_iff.mpr {i : α | f i ≠ 0}.to_finite, },
+  { exact mem_ℓp_infty_iff.mpr (set.finite.bdd_above (set.range (λ (i : α), ∥f i∥)).to_finite) },
+  { casesI nonempty_fintype α, exact mem_ℓp_gen ⟨finset.univ.sum _, has_sum_fintype _⟩ }
+end
+
+variables [fintype α]
+
+/-- The canonical `equiv` between `lp E p ≃ pi_Lp p E` when `E : α → Type u` with `[fintype α]`. -/
+def equiv.lp_pi_Lp : lp E p ≃ pi_Lp p E :=
+{ to_fun := λ f, f,
+  inv_fun := λ f, ⟨f, mem_ℓp.all f⟩,
+  left_inv := λ f, lp.ext $ funext $ λ x, rfl,
+  right_inv := λ f, funext $ λ x, rfl }
+
+lemma coe_equiv_lp_pi_Lp (f : lp E p) : equiv.lp_pi_Lp f = f := rfl
+lemma coe_equiv_lp_pi_Lp_symm (f : pi_Lp p E) : (equiv.lp_pi_Lp.symm f : Π i, E i) = f :=  rfl
+
+lemma equiv_lp_pi_Lp_norm (f : lp E p) : ∥equiv.lp_pi_Lp f∥ = ∥f∥ :=
+begin
+  unfreezingI { rcases p.trichotomy with (rfl | rfl | h) },
+  { rw [pi_Lp.norm_eq_card, lp.norm_eq_card_dsupport], refl },
+  { rw [pi_Lp.norm_eq_csupr, lp.norm_eq_csupr], refl },
+  { rw [pi_Lp.norm_eq_sum h, lp.norm_eq_tsum_rpow h, tsum_fintype], refl },
+end
+
+/-- The canonical `add_equiv` between `lp E p` and `pi_Lp p E` when `E : α → Type u` with
+`[fintype α]` and `[fact (1 ≤ p)]`. -/
+def add_equiv.lp_pi_Lp [fact (1 ≤ p)] : lp E p ≃+ pi_Lp p E :=
+{ map_add' := λ f g, rfl,
+  .. equiv.lp_pi_Lp }
+
+@[simp] lemma coe_add_equiv_lp_pi_Lp [fact (1 ≤ p)] (f : lp E p) :
+  add_equiv.lp_pi_Lp f = f := rfl
+@[simp] lemma coe_add_equiv_lp_pi_Lp_symm [fact (1 ≤ p)] (f : pi_Lp p E) :
+  (add_equiv.lp_pi_Lp.symm f : Π i, E i) = f :=  rfl
```

Comment:
I don't think these should be simp lemmas as you are breaking the barrier between type synonyms.